File: Python/103.binary-tree-zigzag-level-order-traversal.py
#
# @lc app=leetcode id=103 lang=python3
#
# [103] Binary Tree Zigzag Level Order Traversal
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if root == None:
            return None
        result = []
        stack = [root]
        numLevel = 1

        while stack:
            level = []
            stack_len = len(stack)
            for i in range(stack_len):
                node = stack.pop(0)
                level.append(node.val)
                if node.left:
                    stack.append(node.left)
                if node.right:
                    stack.append(node.right)

            if numLevel % 2 == 0:
                level = reversed(level)
            numLevel += 1
            result.append(level)
        
        return result
# @lc code=end

File: Java/103.binary-tree-zigzag-level-order-traversal.java
import java.util.*;

/*
 * @lc app=leetcode id=103 lang=java
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Queue<TreeNode> temp = new LinkedList<>();
        boolean flag = true;
        temp.offer(root);
        while (!temp.isEmpty()) {
            List<Integer> level = new ArrayList<>();
            int len = temp.size();
            if (flag) {
                while (len-- > 0) {
                    TreeNode top = temp.remove();
                    level.add(top.val);
                    if (top.left != null) {
                        temp.offer(top.left);
                    }
                    if (top.right != null) {
                        temp.offer(top.right);
                    }
                }
                flag = false;
            } else {
                while (len-- > 0) {
                    TreeNode top = temp.remove();
                    level.add(0, top.val);
                    if (top.left != null) {
                        temp.offer(top.left);
                    }
                    if (top.right != null) {
                        temp.offer(top.right);
                    }
                }
                flag = true;
            }
            result.add(level);
        }
        return result;
    }
}
// @lc code=end

